howtos/diversification.rst

```diff
@@ -0,0 +1,37 @@
+Diversification of Search Results
+=================================
+
+.. contents:: Table of contents
+
+Introduction
+------------
+
+Xapian allows for diversification of documents which are stored in the form of an MSet.
+This feature is a well-known technique in information retrieval used to increase
+user satisfaction, especially for ambiguous queries.
+
+Xapian currently has an implementation of an *implict* method (using documents as features,
+as opposed to using query based features such as query logs) adapted from the C :sup:`2` - GLS method mentioned in Scalable and Efficient Web Search Results Diversification, Naini et al. 2016. This saves the cost of not having to provide external features such as query
+logs, while still achieving the desired diversification effect, which according to
+the paper is reasonable enough for practical uses as tested on the public data set - ClueWeb09 with TREC Web 09/10 queries.
+
+API
+---
+ 
+Diversification on an MSet of results can be achieved by using the
+:xapian-method:`Diversify` class, e.g.::
+
+    // Query a database and get 10 results, where 'enq' is an instantiated
+    // Enquire object over a database
+    matches = enq.get_mset(0, 10)
```

Comment:
Rather than giving code in a particular language (or here what appears to be Python but with C++ comments!) it's better to use put the example code under `code/` and use `.. xapianexample::` to pull in parts of it.  You can show the results of running the code with `..xapianruinexample::` (see for example `howtos/boolean_filters.rst` for uses of these).

By using these macros, people can provide translations of the examples to other languages, and so we can have a versions of this guide in C++, Python, Perl, PHP, etc.

The other thing these macros do is ensure that the shown example code actually works and produces any output shown.

Reply:
Further, this encourages creating a complete working program; however, you don't have to show all the code in the documentation. It's generally a good idea to base your code on one of the other examples, which you can reference and then describe just what you've changed. In this case, you wouldn't need this snippet of code, which is common in existing search examples.